src/org/opensolaris/opengrok/configuration/messages/Message.java

```diff
@@ -260,135 +176,238 @@ public boolean equals(Object obj) {
                         String.format(
                                 "Message was not accepted by the remote server - "
                                 + "%s (error %#x).",
                                 message.length() > 0 ? message : "too many messages in the system",
-                                c));
+                                deliveryStatus.statusCode));
             default:
                 throw new IOException(
                         String.format(
                                 "Message was not accepted by the remote server %s%s%s(error %#x).",
                                 message.length() > 0 ? "- " : "",
                                 message.length() > 0 ? message : "",
                                 message.length() > 0 ? " " : "",
-                                c));
+                                deliveryStatus.statusCode));
+        }
+    }
+
+    public enum DeliveryStatus {
+
+        OK(0x1), OVER_LIMIT(0x2), ERROR(0x4);
+
+        private final int statusCode;
+
+        DeliveryStatus(final int statusCode) {
+            this.statusCode = statusCode;
+        }
+
+        public static DeliveryStatus fromStatusCode(final int statusCode) {
+            for (DeliveryStatus ds : values()) {
+                if (ds.statusCode == statusCode) {
+                    return ds;
+                }
+            }
+            throw new IllegalArgumentException("Unknown status code " + statusCode);
+        }
+
+        public int getStatusCode() {
+            return statusCode;
+        }
+    }
+
+    private static class MessageDeserializer implements JsonDeserializer<Message> {
+
+        @Override
+        public Message deserialize(
+                final JsonElement jsonElement,
+                final Type type,
+                final JsonDeserializationContext context
+        ) {
+            String classStr = jsonElement.getAsJsonObject().get("type").getAsString();
+            Class<?> cl;
+            try {
+                cl = Class.forName(classStr);
+            } catch (ClassNotFoundException e) {
+                throw new JsonParseException(e);
+            }
+            return context.deserialize(jsonElement, cl);
+        }
+    }
+
+    private static class DurationAdapter extends TypeAdapter<Duration> {
+
+        @Override
+        public void write(final JsonWriter writer, final Duration duration) throws IOException {
+            if (duration == null) {
+                writer.nullValue();
+                return;
+            }
+            writer.value(duration.toString());
+        }
+
+        @Override
+        public Duration read(final JsonReader reader) throws IOException {
+            if (reader.peek() == JsonToken.NULL) {
+                reader.nextNull();
+                return null;
+            }
+            return Duration.parse(reader.nextString());
+        }
+    }
+
+    public static class Builder<T extends Message> {
```

Comment:
I wonder what is the advantage of this builder compared to having a message factory class ?

Reply:
I like this approach more with the builder

Reply:
Well, Google Protobufs use Builder as well and I kind of liked it when I was using those. 

However, I like Builder more because now you could create a lot of new message types. Each a little different and you would not have to have one big factory class. 

However, each have pros and cons. If you prefer factory class I can look into it in more detail.

Reply:
I like the Builder pattern too, just am not aware of its pros and cons. The one-line usage looks nice, as is using reflection to get the constructor which saves space and is more flexible compared to the factory class.